Implement NetSIO/FujiNet support

mozzwald opened the pull request:
These changes implement NetSIO support in Atari800 which can be used to emulate the FujiNet SIO peripheral. 

To test:
 * Use `--enable-netsio` configure flag to enable netsio support when building atari800
 * Use `-netsio` flag when running atari800 to turn on NetSIO
 * Download (or compile) the latest FujiNet-PC binary from [fujinet-firmware](https://github.com/FujiNetWIFI/fujinet-firmware/releases)
 * Run FujiNet-PC from it's directory `cd /path/to/fujinet-pc && ./run-fujinet`
 * The emulator should automatically boot FujiNet CONFIG over NetSIO

Notes:

 * When running atari800 without `-netsio` it should run "as normal". With the flag, *all* SIO traffic is forwarded in/out over the NetSIO connection to FujiNet-PC.
 * SIO Devices tested: Disk, RDevice, Fuji (0x70) and Network (0x71-0x78)
 * Tested working with MacOS and Linux only (Windows unlikely to work and I have no experience or interest compiling for it)
 * Tested working on ARM single board computer dual core 1ghz with 128Mb RAM (Allwinner T113-S3)

mozzwald committed:
pia: add SIO PROCEED and INTERRUPT registers

mozzwald committed:
pia: add SIO PROCEED and INTERRUPT registers

mozzwald committed:
Non working fujinet framework

mozzwald committed:
Non working fujinet framework

mozzwald committed:
Non working fujinet framework

mozzwald committed:
Non working fujinet framework

mozzwald committed:
pia: fix update_PIA_IRQ definition

mozzwald committed:
pia: fix update_PIA_IRQ definition

mozzwald committed:
netsio: implement NetSIO thread framework for FujiNet-PC integration

mozzwald committed:
netsio: implement NetSIO thread framework for FujiNet-PC integration

mozzwald committed:
netsio: implement NetSIO thread framework for FujiNet-PC integration

mozzwald committed:
netsio: implement NetSIO thread framework for FujiNet-PC integration

mozzwald committed:
Add NetSIO commands

mozzwald committed:
Add NetSIO commands

mozzwald committed:
netsio: update handling ping, alive and baud change

mozzwald committed:
netsio: update handling ping, alive and baud change

mozzwald committed:
netsio: update handling ping, alive and baud change

mozzwald committed:
netsio: update handling ping, alive and baud change

mozzwald committed:
netsio: attempt cmd line handling, not working

mozzwald committed:
netsio: attempt cmd line handling, not working

mozzwald committed:
netsio: attempt cmd line handling, not working

mozzwald committed:
netsio: attempt cmd line handling, not working

mozzwald committed:
netsio: updates to command frame handling

mozzwald committed:
netsio: updates to command frame handling

mozzwald committed:
netsio: updates to command frame handling

mozzwald committed:
netsio: updates to command frame handling

mozzwald committed:
netsio: send proper data byte command to fujinet-pc. make emulator wait for fujinet to connect before starting

mozzwald committed:
netsio: send proper data byte command to fujinet-pc. make emulator wait for fujinet to connect before starting

mozzwald committed:
netsio: send proper data byte command to fujinet-pc. make emulator wait for fujinet to connect before starting

mozzwald committed:
netsio: send proper data byte command to fujinet-pc. make emulator wait for fujinet to connect before starting

mozzwald committed:
netsio: disable SIO patch routines if netsio enabled

mozzwald committed:
netsio: disable SIO patch routines if netsio enabled

mozzwald committed:
netsio: disable SIO patch routines if netsio enabled

mozzwald committed:
netsio: disable SIO patch routines if netsio enabled

mozzwald committed:
pokey: allow netsio in non patched mode

mozzwald committed:
pokey: allow netsio in non patched mode

mozzwald committed:
pokey: allow netsio in non patched mode

mozzwald committed:
pokey: allow netsio in non patched mode

mozzwald committed:
netsio: change fujinet rx logic with switch case to handle commands better

mozzwald committed:
netsio: change fujinet rx logic with switch case to handle commands better

mozzwald committed:
sio: try to implement netsio byte handling

mozzwald committed:
sio: try to implement netsio byte handling

mozzwald committed:
sio: try to implement netsio byte handling

mozzwald committed:
sio: try to implement netsio byte handling

mozzwald committed:
netsio: boot from fn-pc disk at regular speed working

mozzwald committed:
netsio: boot from fn-pc disk at regular speed working

mozzwald committed:
netsio: boot from fn-pc disk at regular speed working

mozzwald committed:
netsio: boot from fn-pc disk at regular speed working

a8jan committed:
netsio: boot FujiNet CONFIG

- Handle disk HSIO 0x3F (cheat a bit)

Added some Fuji device commands (for CONFIG):

- Get WiFi Enabled, Get WiFi Status
- Read Host Slots, Write Host Slots
- Read Device Slots
- Get Adapter Config Extended
- Open, Close, Seek, Read Directory

Note: Directory ops are not stable (times out).

a8jan committed:
netsio: boot FujiNet CONFIG

- Handle disk HSIO 0x3F (cheat a bit)

Added some Fuji device commands (for CONFIG):

- Get WiFi Enabled, Get WiFi Status
- Read Host Slots, Write Host Slots
- Read Device Slots
- Get Adapter Config Extended
- Open, Close, Seek, Read Directory

Note: Directory ops are not stable (times out).

a8jan committed:
netsio: boot FujiNet CONFIG

- Handle disk HSIO 0x3F (cheat a bit)

Added some Fuji device commands (for CONFIG):

- Get WiFi Enabled, Get WiFi Status
- Read Host Slots, Write Host Slots
- Read Device Slots
- Get Adapter Config Extended
- Open, Close, Seek, Read Directory

Note: Directory ops are not stable (times out).

a8jan committed:
netsio: boot FujiNet CONFIG

- Handle disk HSIO 0x3F (cheat a bit)

Added some Fuji device commands (for CONFIG):

- Get WiFi Enabled, Get WiFi Status
- Read Host Slots, Write Host Slots
- Read Device Slots
- Get Adapter Config Extended
- Open, Close, Seek, Read Directory

Note: Directory ops are not stable (times out).

mozzwald committed:
Add additional FujiNet commands. Mount and boot disk possible.

mozzwald committed:
Add additional FujiNet commands. Mount and boot disk possible.

mozzwald committed:
Add additional FujiNet commands. Mount and boot disk possible.

mozzwald committed:
Add additional FujiNet commands. Mount and boot disk possible.

mozzwald committed:
Remove unused fujinet testing sources

mozzwald committed:
Remove unused fujinet testing sources

mozzwald committed:
Remove unused fujinet testing sources

mozzwald committed:
Remove unused fujinet testing sources

mozzwald committed:
netsio: make sure it is enabled before calling functions

mozzwald committed:
netsio: make sure it is enabled before calling functions

mozzwald committed:
netsio: make sure it is enabled before calling functions

mozzwald committed:
netsio: make sure it is enabled before calling functions

mozzwald committed:
sio: return true for initialize() or it could fail in atari.c

mozzwald committed:
sio: return true for initialize() or it could fail in atari.c

mozzwald committed:
sio: return true for initialize() or it could fail in atari.c

mozzwald committed:
sio: return true for initialize() or it could fail in atari.c

mozzwald committed:
netsio: increment sync counter before sending, not after

mozzwald committed:
netsio: increment sync counter before sending, not after

mozzwald committed:
Revert "netsio: increment sync counter before sending, not after"

This reverts commit 95e26b62c2960f6708df517423c9844e65efa06a.

mozzwald committed:
Revert "netsio: increment sync counter before sending, not after"

This reverts commit 95e26b62c2960f6708df517423c9844e65efa06a.

mozzwald committed:
sio: handle more fujinet commands

mozzwald committed:
sio: handle more fujinet commands

mozzwald committed:
sio: handle more fujinet commands

mozzwald committed:
sio: handle more fujinet commands

mozzwald committed:
netsio: increment sync counter before sending, not after

mozzwald committed:
netsio: increment sync counter before sending, not after

a8jan committed:
netsio: device and command agnostic

a8jan committed:
netsio: device and command agnostic

a8jan committed:
netsio: device and command agnostic

a8jan committed:
netsio: device and command agnostic

mozzwald committed:
netsio: send cold/warm reset from emu, disable wait for fujinet at startup

mozzwald committed:
netsio: send cold/warm reset from emu, disable wait for fujinet at startup

dillera committed:
Fix SDL include path for macOS compatibility

- Added preprocessor conditionals to handle different SDL header locations
- Use "SDL.h" on macOS systems where SDL headers are in the root include path
- Use "SDL/SDL.h" on other systems (Linux) where headers are in a subdirectory
- This eliminates the need for a symlink and makes the code work on both platforms automatically

dillera committed:
Fix SDL include path for macOS compatibility

- Added preprocessor conditionals to handle different SDL header locations
- Use "SDL.h" on macOS systems where SDL headers are in the root include path
- Use "SDL/SDL.h" on other systems (Linux) where headers are in a subdirectory
- This eliminates the need for a symlink and makes the code work on both platforms automatically

dillera committed:
Fix NetSIO socket compatibility for macOS

This commit addresses multiple BSD socket API compatibility issues on macOS:

1. Added proper handling of the sin_len field in sockaddr_in structures
   (required on macOS but not used on Linux)

2. Fixed socket address length handling to use the correct size for IPv4
   addresses in both send and receive operations

3. Made MSG_NOSIGNAL flag conditional to avoid EINVAL errors on macOS
   while maintaining the functionality on Linux

4. Ensured proper broadcast socket configuration for all platforms

5. Added comprehensive comments explaining platform-specific socket API
   differences to improve maintainability

These changes resolve the EINVAL errors when sending to FujiNet on macOS
while maintaining full compatibility with Linux systems.

dillera committed:
Fix NetSIO socket compatibility for macOS

This commit addresses multiple BSD socket API compatibility issues on macOS:

1. Added proper handling of the sin_len field in sockaddr_in structures
   (required on macOS but not used on Linux)

2. Fixed socket address length handling to use the correct size for IPv4
   addresses in both send and receive operations

3. Made MSG_NOSIGNAL flag conditional to avoid EINVAL errors on macOS
   while maintaining the functionality on Linux

4. Ensured proper broadcast socket configuration for all platforms

5. Added comprehensive comments explaining platform-specific socket API
   differences to improve maintainability

These changes resolve the EINVAL errors when sending to FujiNet on macOS
while maintaining full compatibility with Linux systems.

mozzwald committed:
netsio: add configure option to enable netsio. cleanup. add ifdefs for netsio

mozzwald committed:
netsio: add configure option to enable netsio. cleanup. add ifdefs for netsio

mozzwald committed:
netsio: add configure option to enable netsio. cleanup. add ifdefs for netsio

mozzwald committed:
netsio: add configure option to enable netsio. cleanup. add ifdefs for netsio

mozzwald committed:
netsio: fix typo for log_print

mozzwald committed:
netsio: fix typo for log_print

mikrosk commented:
Thanks! A couple of notes before I look into that:

> When running atari800 without -netsio it should run "as normal". With the flag, all SIO traffic is forwarded in/out over the NetSIO connection to FujiNet-PC.

My physical FujiNet cartridge has an additional SIO port which (I believe) allows forwarding SIO traffic and therefore connecting normal peripherals as well. This functionality is not desired in your emulated code? I wonder how would one e.g. boot something from FujiNet (e.g. a DOS) but wanted to save files to his local D2: ATR image.

> Tested working with MacOS and Linux only (Windows unlikely to work and I have no experience or interest compiling for it)

While there are few features not available on other platforms, you should at least ensure that your code is properly `#ifdef`'ed, i.e. that it a) prints an error message that given feature is not possible to compile on Windows  (we are using MSYS, i.e. cross compiling from Linux anyway) and b) Atari800 is still compilable with your code changes.

Oh and please rebase your branch on latest master, so we are sure there are no conflicts pending.

mozzwald commented:
> My physical FujiNet cartridge has an additional SIO port which (I believe) allows forwarding SIO traffic and therefore connecting normal peripherals as well. This functionality is not desired in your emulated code? I wonder how would one e.g. boot something from FujiNet (e.g. a DOS) but wanted to save files to his local D2: ATR image.

Indeed it does. We went for the path of least resistance here and FujiNet can replace many of the built in atari800 devices so one can use FujiNet instead of the internal emulator devices. FujiNet-PC has a virtual SD card for local storage. You can symlink that to some other dir on your computer that contains disk images if you like.

> While there are few features not available on other platforms, you should at least ensure that your code is properly `#ifdef`'ed, i.e. that it a) prints an error message that given feature is not possible to compile on Windows (we are using MSYS, i.e. cross compiling from Linux anyway) and b) Atari800 is still compilable with your code changes.

I will look into how I can disable building NetSIO for windows. If you know the best approach to do that, please let me know.

mikrosk commented:
> I will look into how I can disable building NetSIO for windows. If you know the best approach to do that, please let me know.

Take a look at e.g. R: serial device handling (Unix/Linux only) in `configure.ac`, `rserial` and `R_SERIAL` variables should give you some hints.

mozzwald committed:
netsio: do not allow on windows build

mozzwald committed:
netsio: do not allow on windows build

mozzwald commented:
I pushed a change that should only allow building netsio for linux/mac

mikrosk commented:
@mozzwald please rebase it again, it turned out that even our master branch couldn't build for some time. :-P

mozzwald committed:
netsio: remove old testing if statement

mozzwald committed:
netsio: remove old testing if statement

dillera commented:
The day has finally come.

amiga23 commented:
Do you need feedback before merging? Works for me, tested on Manjaro Linux.
`./configure --with-sdl --with-opengl --with-readline --with-mp3=lame --with-sound=sdl --with-video=sdl --enable-netsio`

dillera commented:
What is stopping this merge? Wondering what we can do here.

mikrosk commented:
@dillera at least some sort of code review has to be done. With our limited resources I can't promise you an immediate attention but we'll do our best to review/merge it ASAP as it is a very nice feature to be had.

mikrosk commented on src/sio.c:
```diff
@@ -171,7 +174,8 @@ char SIO_status[256];
 static UBYTE CommandFrame[6];
 static int CommandIndex = 0;
-static UBYTE DataBuffer[256 + 3];
+/*static UBYTE DataBuffer[256 + 3];*/
```
Is this intentional to keep? I mean, if the 64k buffer below is enough (and there are no side effects, like using it as a ring buffer or something), this doesn't have to be here.

mozzwald replied:
This can be removed

mikrosk commented on src/pokey.c:
```diff
@@ -259,6 +262,11 @@ void POKEY_PutByte(UWORD addr, UBYTE byte)
 		if ((POKEY_SKCTL & 0x70) == 0x20 && POKEY_siocheck())
 			SIO_PutByte(byte);
+#ifdef NETSIO
+		else if (netsio_enabled && (POKEY_SKCTL & 0x70) == 0x70) /* TODO: proper way to enable modem */
```
Do you plan to work on this modem feature later?

mozzwald replied:
@a8jan added this comment. I did test modem with BobTerm and it worked so not sure what is meant here

a8jan replied:
When I was testing various FujiNet provided peripherals with atari800, I've noticed modem was not working.
Quick fix was to test `(POKEY_SKCTL & 0x70) == 0x70` instead of calling more complex `POKEY_siocheck()`.
Modem started to work (tested Ice-T with CP/M, BobTerm). However, I'm not sure how to test POKEY registers for only valid serial port output mode...

mikrosk replied:
@a8jan please add this comment to the code, it will be beneficial for anyone attempting to provide a proper fix

mikrosk commented on src/pokey.c:
```diff
@@ -500,6 +508,19 @@ void POKEY_Scanline(void)
+	if (netsio_enabled && POKEY_DELAYED_SERIN_IRQ == 0) {
+		int avail = netsio_available();
+		if (avail > 0) {
+			 /* TODO make various SIO speeds working, at least 19200 ;-) */
```
What prevents more SIO speeds working now?

mozzwald replied:
I think we need to figure out out the proper calculations for POKEY_DELAYED_SERIN_IRQ here, but I'm not sure. Maybe @a8jan can comment on this

a8jan replied:
Hello / Zdravím,
Yes, currently the POKEY_DELAYED_SERIN_IRQ is set to the same values ignoring the actual speed.
"at least 19200" is from the time when I started with much higher values in POKEY_DELAYED_SERIN_IRQ (lower speed) to get it working. This part of the comment can be removed.

mikrosk replied:
@a8jan same as above, please adjust/extend the comment.

mikrosk reviewed (commented):
Looks very good, thanks for keeping the code base so clean. I have a few of questions:
- We are considering making a new release soon. Do you feel confident enough to include (enable) this feature in the release build?
- Are you planning on maintaining / updating this feature in the future or is this more or less a one-time contribution?
- When merging, is it ok to merge as one commit or would you prefer some parts separated?

mikrosk commented:
A couple of warning with netsio both enabled and disabled:
```
pia.c:127:13: warning: ‘set_CB1’ defined but not used [-Wunused-function]
  127 | static void set_CB1(int value)
      |             ^~~~~~~
pia.c:116:13: warning: ‘set_CA1’ defined but not used [-Wunused-function]
  116 | static void set_CA1(int value)
      |             ^~~~~~~
```
A few warnings with netsio enabled:
```
  CC       atari800-netsio.o
netsio.c:54:7: warning: no previous prototype for ‘buf_to_hex’ [-Wmissing-prototypes]
   54 | char *buf_to_hex(const uint8_t *buf, size_t offset, size_t len) {
      |       ^~~~~~~~~~
netsio.c: In function ‘buf_to_hex’:
netsio.c:60:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   60 |     char *p = s;
      |     ^~~~
netsio.c: In function ‘send_to_fujinet’:
netsio.c:108:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  108 |     socklen_t addr_len = sizeof(struct sockaddr_in);
      |     ^~~~~~~~~
netsio.c:140:18: warning: unused variable ‘errmsg’ [-Wunused-variable]
  140 |             char errmsg[256];
      |                  ^~~~~~
netsio.c: At top level:
netsio.c:176:6: warning: no previous prototype for ‘send_byte_to_fujinet’ [-Wmissing-prototypes]
  176 | void send_byte_to_fujinet(uint8_t data_byte) {
      |      ^~~~~~~~~~~~~~~~~~~~
netsio.c:184:6: warning: no previous prototype for ‘send_block_to_fujinet’ [-Wmissing-prototypes]
  184 | void send_block_to_fujinet(const uint8_t *block, size_t len) {
      |      ^~~~~~~~~~~~~~~~~~~~~
netsio.c: In function ‘send_block_to_fujinet’:
netsio.c:187:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  187 |     uint8_t packet[512 + 2];
      |     ^~~~~~~
netsio.c: In function ‘netsio_init’:
netsio.c:241:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  241 |     int broadcast = 1;
      |     ^~~
netsio.c:202:26: warning: unused variable ‘tx_thread’ [-Wunused-variable]
  202 |     pthread_t rx_thread, tx_thread;
      |                          ^~~~~~~~~
netsio.c: In function ‘netsio_cmd_on’:
netsio.c:311:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  311 |     uint8_t p = NETSIO_COMMAND_ON;
      |     ^~~~~~~
netsio.c: In function ‘netsio_cmd_off_sync’:
netsio.c:334:47: warning: initializer element is not computable at load time [-Wpedantic]
  334 |     uint8_t p[2] = { NETSIO_COMMAND_OFF_SYNC, netsio_sync_num };
      |                                               ^~~~~~~~~~~~~~~
netsio.c:334:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  334 |     uint8_t p[2] = { NETSIO_COMMAND_OFF_SYNC, netsio_sync_num };
      |     ^~~~~~~
netsio.c:335:21: warning: passing argument 1 of ‘send_to_fujinet’ from incompatible pointer type [-Wincompatible-pointer-types]
  335 |     send_to_fujinet(&p, sizeof(p));
      |                     ^~
      |                     |
      |                     uint8_t (*)[2] {aka unsigned char (*)[2]}
netsio.c:93:44: note: expected ‘const uint8_t *’ {aka ‘const unsigned char *’} but argument is of type ‘uint8_t (*)[2]’ {aka ‘unsigned char (*)[2]’}
   93 | static void send_to_fujinet(const uint8_t *pkt, size_t len) {
      |                             ~~~~~~~~~~~~~~~^~~
netsio.c: In function ‘netsio_send_byte’:
netsio.c:351:42: warning: initializer element is not computable at load time [-Wpedantic]
  351 |     uint8_t pkt[2] = { NETSIO_DATA_BYTE, b };
      |                                          ^
netsio.c:355:21: warning: passing argument 1 of ‘send_to_fujinet’ from incompatible pointer type [-Wincompatible-pointer-types]
  355 |     send_to_fujinet(&pkt, 2);
      |                     ^~~~
      |                     |
      |                     uint8_t (*)[2] {aka unsigned char (*)[2]}
netsio.c:93:44: note: expected ‘const uint8_t *’ {aka ‘const unsigned char *’} but argument is of type ‘uint8_t (*)[2]’ {aka ‘unsigned char (*)[2]’}
   93 | static void send_to_fujinet(const uint8_t *pkt, size_t len) {
      |                             ~~~~~~~~~~~~~~~^~~
netsio.c: In function ‘netsio_send_byte_sync’:
netsio.c:374:45: warning: initializer element is not computable at load time [-Wpedantic]
  374 |     uint8_t p[3] = { NETSIO_DATA_BYTE_SYNC, b, netsio_sync_num};
      |                                             ^
netsio.c:374:48: warning: initializer element is not computable at load time [-Wpedantic]
  374 |     uint8_t p[3] = { NETSIO_DATA_BYTE_SYNC, b, netsio_sync_num};
      |                                                ^~~~~~~~~~~~~~~
netsio.c:374:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  374 |     uint8_t p[3] = { NETSIO_DATA_BYTE_SYNC, b, netsio_sync_num};
      |     ^~~~~~~
netsio.c:375:21: warning: passing argument 1 of ‘send_to_fujinet’ from incompatible pointer type [-Wincompatible-pointer-types]
  375 |     send_to_fujinet(&p, sizeof(p));
      |                     ^~
      |                     |
      |                     uint8_t (*)[3] {aka unsigned char (*)[3]}
netsio.c:93:44: note: expected ‘const uint8_t *’ {aka ‘const unsigned char *’} but argument is of type ‘uint8_t (*)[3]’ {aka ‘unsigned char (*)[3]’}
   93 | static void send_to_fujinet(const uint8_t *pkt, size_t len) {
      |                             ~~~~~~~~~~~~~~~^~~
netsio.c: In function ‘fujinet_rx_thread’:
netsio.c:453:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  453 |         ssize_t n = recvfrom(sockfd,
      |         ^~~~~~~
netsio.c:483:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  483 |         uint8_t cmd = buf[0];
      |         ^~~~~~~
netsio.c:534:17: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  534 |                 uint8_t reply[2] = { NETSIO_CREDIT_UPDATE, 3 };
      |                 ^~~~~~~
netsio.c:552:17: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  552 |                 uint32_t baud = buf[1]
      |                 ^~~~~~~~
netsio.c:552:26: warning: unused variable ‘baud’ [-Wunused-variable]
  552 |                 uint32_t baud = buf[1]
      |                          ^~~~
netsio.c:574:17: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  574 |                 uint8_t  resp_sync  = buf[1];
      |                 ^~~~~~~
netsio.c:642:17: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  642 |                 uint8_t data = buf[1];
      |                 ^~~~~~~
netsio.c:661:17: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  661 |                 size_t payload_len = n - 1;
      |                 ^~~~~~
netsio.c:431:13: warning: unused variable ‘packet’ [-Wunused-variable]
  431 |     uint8_t packet[65536];
      |             ^~~~~~
netsio.c: At top level:
netsio.c:52:14: warning: ‘emu_tx_thread’ declared ‘static’ but never defined [-Wunused-function]
   52 | static void *emu_tx_thread(void *arg);
      |              ^~~~~~~~~~~~~
netsio.c: In function ‘netsio_send_block’:
netsio.c:365:1: warning: control reaches end of non-void function [-Wreturn-type]
  365 | }
      | ^
```

Btw what is the difference between "FujiNet" and "NetSIO" ?

mozzwald commented:
> * We are considering making a new release soon. Do you feel confident enough to include (enable) this feature in the release build?

I feel it's working well enough to include. I've been using it almost daily to test some apps I'm working on for FujiNet.

> * Are you planning on maintaining / updating this feature in the future or is this more or less a one-time contribution?

I'm sure we will continue to improve the emulation as time goes on and can continue to submit PRs.

> * When merging, is it ok to merge as one commit or would you prefer some parts separated?

I think a single commit is fine. The reason for the large amount was due to multiple people working on it and needing to share the current state.

mozzwald commented:
> Btw what is the difference between "FujiNet" and "NetSIO" ?

FujiNet is the SIO Peripheral emulator which began as a real physical hardware device. NetSIO is a protocol to pass SIO communication over a network (a bridge if you will) created to facilitate using FujiNet-PC (FujiNet "firmware" running on a PC) with emulators.

I can understand the confusion and had wondered if it best to call this FujiNet instead of NetSIO in atari800. Technically, NetSIO could be used for other purposes so I went with NetSIO. I'm not really tied to one or the other, so whatever makes more sense to name it. Perhaps leave the code intact as netsio and just rename the enable flags to `--enable-fujinet` and `-fujinet` :shrug:

amiga23 commented:
From my point of view as a user, I would expect the full fujinet emulation being part of Atari800, if you calll it fujinet. So for me netsio is better.

mozzwald committed:
netsio: fix compiler warnings

mozzwald committed:
netsio: fix compiler warnings

dillera commented:
> From my point of view as a user, I would expect the full fujinet emulation being part of Atari800, if you calll it fujinet. So for me netsio is better.

Yes, this is an important element to keep clear.

First I agree it should be --netsio and not --fujinet.

The fujinet features/capabilities are _not_ included in any of this code- those require building and running the FujiNet-PC (a posix compatible instance of the physical FujiNet device). As Moz pointed out.

The additional code implements a mechanism to intercept and redirect SIO calls from the running emulation and send them over the network using UDP to a waiting FujiNet-PC process. The SIO frames are encapsulated into "netsio" packets which the Fujinet-PC application understands.

NetSIO is documented here:
https://github.com/FujiNetWIFI/fujinet-firmware/wiki/FEP-003


For platforms that atari800 supports, if they also have a working IP network stack and can send and receive UDP packets, they can participate in this new feature. If they do not have this ability they cannot use this feature.

I know I'm explaining it to people here who probably already understand this, I'm just trying to clarify my own thoughts and if others are reading this who are not as deeply embedded in the FujiNet project- it may help set some context.

mikrosk committed:
Update Windows build instructions

mikrosk commented:
@mozzwald one more observation: I tried to compile it on Windows (namely via MSYS2) and netsio is still compiled in (and failing on `#include <sys/socket.h>`).

Now that I studied your `configure.ac` changes more closely, I see what's wrong. Take the R: serial device (which I had mentioned earlier):
```
if [[ "$a8_host" != "win" -a "$a8_host" != "macos" ]]; then
            A8_OPTION(rserial,yes,
                     [Use the host serial port with the R: networking device (Linux/Unix only) (default=ON)],
                     R_SERIAL,[Define to use the host serial port with the R: device.]
                    )
        else 
            if [[ "$a8_host" = "win" ]]; then
                A8_NEED_LIB(ws2_32)
            fi
            WANT_R_SERIAL="no"
        fi
```
So we check for "win" and "macos" and if the host is one of those, `WANT_R_SERIAL` is explicitly set to "no" and it is not even shown in the list of enabled features after `./configure` has finished. However your code:
```
dnl NetSIO / FujiNet support (only on Linux or MacOS)
    A8_OPTION(netsio,no,
            [Support NetSIO Protocol I/O interface (FujiNet) (default=OFF)],
            NETSIO,[Define to enable NetSIO Protocol/FujiNet support.]
            )
AM_CONDITIONAL([WANT_NETSIO], test "$WANT_NETSIO" = "yes")
if [[ "$a8_host" = "win" ]]; then
    AC_CHECK_HEADERS([windows.h winsock2.h],,WANT_NETSIO=no)
else
    AC_CHECK_HEADERS([arpa/inet.h netdb.h netinet/in.h sys/socket.h termios.h],,WANT_NETSIO=no)
fi
```
just checks for `windows.h` and `winsock2.h` and if found, enables the feature.

So this needs to be fixed.

Apart from that and pending answers from @a8jan there's nothing stopping the merge.

mozzwald committed:
netsio: prevent segfault and run normally if unable to bind netsio socket

mozzwald committed:
netsio: prevent segfault and run normally if unable to bind netsio socket

mozzwald committed:
netsio: fix configure to skip netsio for windows

mozzwald committed:
netsio: fix configure to skip netsio for windows

mozzwald commented:
> So this needs to be fixed.

I pushed a change for the configure. Hopefully that does it. 

Also, a bug was reported today which was also fixed. If atari800 was already running netsio (or perhaps some other app had control of the port) if you run another instance of the emulator with netsio enabled it would segfault. The fix continues emulation with netsio disabled if it cannot open the port.

mikrosk commented on configure.ac:
```diff
@@ -190,11 +191,13 @@ if [[ "$a8_host" = "falcon" ]]; then
 	CFLAGS="-m68020-60 -O2 -fomit-frame-pointer -Wall"
 	LDFLAGS="-m68020-60"
 	SUPPORTS_RDEVICE=no
+    SUPPORTS_NETSIO=no
```
Indentation (perhaps just make the whole block aligned with spaces, not sure where the tabs come from).

mikrosk commented on configure.ac:
```diff
@@ -190,11 +191,13 @@ if [[ "$a8_host" = "falcon" ]]; then
 	CFLAGS="-mcpu=5475 -O2 -fomit-frame-pointer -Wall"
 	LDFLAGS="-mcpu=5475"
 	SUPPORTS_RDEVICE=no
+    SUPPORTS_NETSIO=no
```
Indentation (perhaps just make the whole block aligned with spaces, not sure where the tabs come from).

mikrosk commented on configure.ac:
```diff
@@ -914,16 +917,13 @@ A8_OPTION(pbi_bb,yes,
-            [Support NetSIO Protocol I/O interface (FujiNet) (default=OFF)],
-            NETSIO,[Define to enable NetSIO Protocol/FujiNet support.]
-            )
+A8_OPTION(netsio,no,
```
Indentation (at some places `dnl` is aligned with `if` but AFAICS `A8_OPTION` is always indented).

mikrosk commented on configure.ac:
```diff
@@ -914,16 +917,13 @@ A8_OPTION(pbi_bb,yes,
-            NETSIO,[Define to enable NetSIO Protocol/FujiNet support.]
-            )
+A8_OPTION(netsio,no,
+        [Support NetSIO Protocol I/O interface (FujiNet) (default=OFF)],
```
Since you have approved this feature for the release, I think it should be ON by default (so packagers don't overlook its presence).

mikrosk commented on configure.ac:
```diff
@@ -1418,7 +1418,11 @@ echo "Using MIO emulation?..................: $WANT_PBI_MIO"
 echo "Using IDE emulation?..................: $WANT_IDE"
 echo "Using Pokey registers recording?......: $WANT_POKEYREC"
-echo "Using NetSIO/FujiNet emulation?.......: $WANT_NETSIO"
+
```
Perhaps the additional empty line(s) are not needed here.

mikrosk replied:
Sorry, I should have been more specific -- also the `fi` line doesn't need an empty line (aligned with other features).

mikrosk replied:
@mozzwald when fixing the windows detection, please fix also this ultra-minor problem, i.e. remove the empty line after `fi`.

mozzwald replied:
> @mozzwald when fixing the windows detection, please fix also this ultra-minor problem, i.e. remove the empty line after `fi`.

done

mikrosk commented on configure.ac:
```diff
@@ -351,9 +354,9 @@ AC_HEADER_TIOCGWINSZ
 AC_CHECK_HEADERS([fcntl.h sys/ioctl.h sys/soundcard.h],,SUPPORTS_SOUND_OSS=no)
 if [[ "$a8_host" = "win" ]]; then
-    AC_CHECK_HEADERS([windows.h winsock2.h],,SUPPORTS_RDEVICE=no)
+    AC_CHECK_HEADERS([windows.h winsock2.h],,SUPPORTS_RDEVICE=no; SUPPORTS_NETSIO=no)
```
This check alone is still not enough. Even checking for R: serial device is questionable because we don't support it at all on Windows but perhaps this was for a future implementation (where the serial device, same as NetSIO, would need to use win sockets).

However you are still missing the crucial part which is done for R: device, as mentioned in https://github.com/atari800/atari800/pull/245#issuecomment-2890987673. You have to explicitly disable NetSIO if windows platform is detected (i.e. if SUPPORTS_NETSIO then if windows then WANT_NETSIO="no")

mikrosk replied:
Sorry, still not there. Regardless of `--enable-netsio` / `--disable-netsio` I'm getting on MSYS2:

```
configure: error: conditional "WANT_NETSIO" was never defined.
```
If you don't want to shoot in the dark, you can install e.g. https://github.com/HolyBlackCat/quasi-msys2 for debugging Window issues.

mozzwald replied:
> @mozzwald this one still needs your attention. :)

Can you throw me a bone here and just give me the fix? I'm have zero interest in installing msys2 to test this.

dillera committed:
fixing up configure.ac for request 245 in the PR

dillera committed:
fixing up configure.ac for request 245 in the PR

mozzwald committed:
Merge pull request #4 from dillera/netsio-device-agnostic

fixing up configure.ac for request 245 in the PR

mozzwald commented:
@mikrosk new commit added to hopefully fix those issues

mikrosk commented:
Still a few warnings to kill (reported by gcc 15.1.1), excluding the `set_C?1` ones:
```
  CC       atari800-sio.o
../../atari800.git/src/sio.c: In function ‘NetSIO_GetByte’:
../../atari800.git/src/sio.c:1715:13: warning: unused variable ‘ts’ [-Wunused-variable]
 1715 |         int ts = TransferStatus;
      |             ^~
```
[...]
```
../../atari800.git/src/netsio.c: In function ‘fujinet_rx_thread’:
../../atari800.git/src/netsio.c:453:13: warning: unused variable ‘packet’ [-Wunused-variable]
  453 |     uint8_t packet[65536];
      |             ^~~~~~
../../atari800.git/src/netsio.c: At top level:
../../atari800.git/src/netsio.c:57:14: warning: ‘emu_tx_thread’ declared ‘static’ but never defined [-Wunused-function]
   57 | static void *emu_tx_thread(void *arg);
      |              ^~~~~~~~~~~~~
../../atari800.git/src/netsio.c:183:13: warning: ‘send_byte_to_fujinet’ defined but not used [-Wunused-function]
  183 | static void send_byte_to_fujinet(uint8_t data_byte) {
      |             ^~~~~~~~~~~~~~~~~~~~
../../atari800.git/src/netsio.c:59:7: warning: ‘buf_to_hex’ defined but not used [-Wunused-function]
   59 | char *buf_to_hex(const uint8_t *buf, size_t offset, size_t len) {
      |       ^~~~~~~~~~

```
I think the `set_C?1` functions should be at least covered within the netsio #ifdef so the warnings don't show when configuring with `--disable-netsio` at least. Or maybe they should be just commented out if you still consider them useful in the future.

Furthermore, there are still those two comments to be added by @a8jan to clarify the TODOs.

mozzwald committed:
netsio: fix warnings, add netsio help msg, fix comments, use custom delay instead of SDL delay in netsio.c

mozzwald committed:
netsio: fix warnings, add netsio help msg, fix comments, use custom delay instead of SDL delay in netsio.c

mozzwald commented:
@mikrosk sorry for the delay

I think this is everything you requested :crossed_fingers:

mozzwald committed:
configure.ac: remove empty line

mozzwald committed:
configure.ac: remove empty line

mozzwald committed:
build.yml: update pkg list before install pkgs

mozzwald committed:
CI: add windows build (#247)

CI: fix linux build with apt update before install

CI: add package depends for windows

CI: fix package names for windows

CI: change windows build name

mozzwald committed:
Merge branch 'netsio-device-agnostic' of github.com:mozzwald/atari800 into netsio-device-agnostic

mozzwald committed:
Fix configure for windows and guard netsio.c with ifdef

mozzwald commented:
How's that? 
:window: :heavy_check_mark:

mikrosk commented:
Congratulations, your PR has been squash-merged in commit https://github.com/atari800/atari800/commit/f6f5f8839944abf95c56f8fbdbed79f62cbda7dd.

When I looked at your latest changes I realised what was wrong with your original WANT_NETSIO detection: you mistakenly put the `AM_CONDITIONAL([WANT_NETSIO], test "$WANT_NETSIO" = "yes")` line within `if` / `fi` statement while it should have been global from the beginning. When fixing that, all the other changes were not needed.

Thank you for your contribution (also for the msys build bot), I'll do my best to accept your further netsio changes at a much faster rate.
